enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,553 @@
+
+In the first phase of the transition all cloud providers will still rely on the in-tree storage. `--external-cloud-volume-plugin=<some-provider>` set on `KCM` pod will preserve the in-tree storage controller loops.
+
+At the later phases, CSI support and CSI migration will be responsible for removing this flag from the `KCM`.
+
+With the `--external-cloud-volume-plugin` flag set, existing designs for this transition are being preserved. The storage will temporarily become a lone in-tree controller depending on the cloud, and could be excluded by setting this flag off. This means that `kcm` will still initialize the `cloudprovider.Interface`, but this will only be used by storage controllers.
+
+#### Pre-release / Development Feature Gate
+
+In order to achieve desired cluster configuration for cloud components during development and technical preview, we will add a new OpenShift [FeatureGate](https://docs.openshift.com/container-platform/4.5/nodes/clusters/nodes-cluster-enabling-features.html) resource named `ExternalCloudProvider`, which will allow developers and customers to force their cluster to upgrade from in-tree to out-of-tree cloud providers.
+
+This feature gate will be shipped with `TechPreviewNoUpgrade` `FeatureSet`, preventing upgrades.
+
+The feature gate will be used to determine whether or not operators should migrate their components to the out-of-tree configuration.
+If the feature gate is found, the operators for components such as `kube-apiserver`, `kube-controller-manager` and `kubelet` will update their component's configuration accordingly to allow hosting external `CCM` for current platform.
+
+Once an out-of-tree provider implementation is ready to be released (GA), it will no longer depend on the feature gate, allowing production clusters to automatically upgrade to the out-of-tree implementation during 4.N to 4.N+1 upgrade.
+
+Once all providers are moved out-of-tree, the feature gate will be safe to remove.
+
+This feature gate will be used in CI tasks, to test migration stability.
+
+#### Migration procedure
+
+The process of the migration will be managed by corresponding operators. Communication will be done via `CloudControllerOwner` condition on the operator's `ClusterOperator` resource.
+
+```yaml
+conditions:
+- type: CloudControllerOwner
+  status: True
+```
+
+`status` will indicate the state of the ownership for this code. Setting it to `False` on one `ClusterOperator` will unblock resource provisioning for the other operator.
+
+The plan for migration is the following:
+
+1) `kubelet` and `kube-controller-manager`.
+    1) Set the `CloudControllerOwner` condition on the `ClusterOperator` with `status=True`.
+    2) Disable cloud loops in its operands.
+    3) Wait for the new resource version rollout to complete and become stable.
+    4) Indicate the completion of the migration by setting `status=False` on the `CloudControllerOwner` condition.
```

Comment:
What if a new node joins the cluster later? Can it run an unconfigured kubelet, i.e. with in-tree cloud provider for some short time? Will something turn Kubelet/MCO `ClusterOperator` to `status=True`?

Reply:
It can. Initialized kubelet will set the taint on the Node object, which `CCM` will be responsible to remove after. Only MCO can set this condition, and it will be healthy at the moment to report `status=True`.